src/CQRS/LeanCode.CQRS.ScopedTransaction/ScopedTransactionMiddleware.cs

```diff
@@ -0,0 +1,89 @@
+using System.Diagnostics.CodeAnalysis;
+using LeanCode.CQRS.Execution;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.EntityFrameworkCore;
+
+namespace LeanCode.CQRS.ScopedTransaction;
+
+public class ScopedTransactionMiddleware<TDbContext>
+    where TDbContext : DbContext
+{
+    private readonly RequestDelegate next;
+
+    public ScopedTransactionMiddleware(RequestDelegate next)
+    {
+        this.next = next;
+    }
+
+    public Task InvokeAsync(HttpContext httpContext, TDbContext dbContext)
+    {
+        return UsePessimisticConcurrency(httpContext, out var config)
+            ? RunWithPessimisticConcurrency(httpContext, dbContext, config)
+            : RunWithOptimisticConcurrency(httpContext, dbContext);
+    }
+
+    private bool UsePessimisticConcurrency(
+        HttpContext httpContext,
+        [NotNullWhen(true)] out PessimisticConcurrencyAttribute? config
+    )
+    {
+        var cqrsMetadata = httpContext.GetCQRSEndpoint().ObjectMetadata;
+        var handlerType = cqrsMetadata.HandlerType;
+
+        if (
+            Attribute.GetCustomAttribute(handlerType, typeof(PessimisticConcurrencyAttribute))
```

Comment:
How is the performance of this? Should we plan on having some kind of cache?

Reply:
Let's put that on our backlog, with more systemic approach (I really hope that this approach will be used thoroughly :D).